### ISSUE TITLE
Support for maintenance mode profile configuration 

### ISSUE THREAD

**Issue by kmueller68** [aristanetworks/ansible-avd#922]
**Issue Type**

- [ ] Template enhancement
- [x] Role enhancement
- [ ] Documentation enhancement (avd.sh)

**Is your feature request related to a problem? Please describe.**

We have to adjust the maintenance mode profile configuration on all L3/MLAG-Leaf switches to make maintenance mode work in our environment.

**Describe the solution you'd like**

We have to configure a maintenance profile interface with certain values like the following one and make it the default.

```
maintenance
   profile interface int_maint default
   !
   profile interface int_maint
      rate-monitoring load-interval 100
      rate-monitoring threshold 1000
      shutdown max-delay 100
```

There is currently no template available within the eos_cli_config_gen role.
The data model could perhaps look like this:
```
maintenance:
  profile:
    interface:
      < profile_interface_1 >:
        rate_monitoring:
          load_interval: < seconds >
          threshold: < kbps >
        shutdown:
          max_delay: < seconds >     
        default: < true | false >
```

**Describe alternatives you've considered**

The current alternative is, to create a config_let within CVP and bind it to the containers of our leafs manually within CVP or "automated" by adding the config_let via AVD.
Our goal is that Ansible code is the single source of configuration.  

**Additional context**

**Comment by daanvdsanden** [comment]
I would like to suggest a different data model:

1. move the default out of the profiles (no need to loop all the profiles to get the default, also what behavior do you want when two profiles have set the default to true).
2. flatten the profiles to have it match the config hierarchy, this also makes the plural form a bit more logical

This makes the model match the cli config hierarchy better and less logic is needed (which can be embedded in the design role). 
```
maintenance:
  default_interface_profile: <interface_profile_1>
  default_bgp_profile: < bgp_profile_1 >
  default_unit_profile: < unit_profile_1 >
  interface_profiles:
    < interface_profile_1 >:
      rate_monitoring:
        load_interval: < seconds >
        threshold: < kbps >
      shutdown:
        max_delay: < seconds >     
  bgp_profiles:
    < bgp_profile_1 >:
      initiator:
        route_map_inout: < route_map >
  unit_profiles:
    < unit_profile_1 >:
      on_boot:
        duration: < 300-3600 >
```
**EDIT:** changed unit and bgp profiles to be more consistent with the rest of the template.

**Comment by ClausHolbechArista** [comment]
Looking good so far. If anyone is up for it, please add `unit`.
Update: Removed `group` since it is a global config. So strictly speaking, it should not be part of `maintenance:`.

**Comment by daanvdsanden** [comment]
```yaml
maintenance:
  default_interface_profile: <interface_profile_1>
  default_bgp_profile: < bgp_profile_1 >
  default_unit_profile: < unit_profile_1 >
  interface_profiles:
    < interface_profile_1 >:
      rate_monitoring:
        load_interval: < seconds >
        threshold: < kbps >
      shutdown:
        max_delay: < seconds >     
  bgp_profiles:
    < bgp_profile_1 >:
      initiator:
        route_map_inout: < route_map >
  unit_profiles:
    < unit_profile_1 >:
      on_boot:
        duration: < 300-3600 >
  units:
    < unit_name_1 >:
      profile: < unit_profile_1 >
      groups:
        bgp_groups:
        - < bgp_group_1>
        - < bgp_group_2>
        interface_groups:
        - < interface_group_1>
        - < interface_group_2>
```
I had a quick start two weeks ago:
daanvdsanden/ansible-avd@43b92bdd3172aa59ab87a5ab8e6138f772baeb79

However looking at the generated config for units, I'm having a hard time figuring out the order in which the added groups appear:
```
vEOS-LF001#conf t
vEOS-LF001(config)#mai
vEOS-LF001(config-maintenance)#unit UNIT2
vEOS-LF001(config-unit-UNIT2)#show active
maintenance
   unit UNIT2
vEOS-LF001(config-unit-UNIT2)#profile unit TEST
vEOS-LF001(config-unit-UNIT2)#group bgp 1_BGP
vEOS-LF001(config-unit-UNIT2)#group bgp 2_BGP
vEOS-LF001(config-unit-UNIT2)#show active
maintenance
   unit UNIT2
      group bgp 2_BGP
      group bgp 1_BGP
      profile unit TEST
vEOS-LF001(config-unit-UNIT2)#group bgp BGP_1
vEOS-LF001(config-unit-UNIT2)#group bgp BGP_2
vEOS-LF001(config-unit-UNIT2)#show active
maintenance
   unit UNIT2
      group bgp 2_BGP
      group bgp BGP_1
      group bgp 1_BGP
      group bgp BGP_2
      profile unit TEST
vEOS-LF001(config-unit-UNIT2)#
```
When I add interface groups it even gets harder to find the order (at least for me):
```
vEOS-LF001(config-unit-UNIT2)#show active
maintenance
   unit UNIT2
      group bgp 2_BGP
      group bgp BGP_1
      group bgp 1_BGP
      group bgp BGP_2
      profile unit TEST
vEOS-LF001(config-unit-UNIT2)#group interface IG_1
vEOS-LF001(config-unit-UNIT2)#group interface IG_2
vEOS-LF001(config-unit-UNIT2)#show active
maintenance
   unit UNIT2
      group interface IG_1
      group bgp 2_BGP
      group bgp BGP_1
      group bgp 1_BGP
      group interface IG_2
      group bgp BGP_2
      profile unit TEST
vEOS-LF001(config-unit-UNIT2)#group interface 1_IG
vEOS-LF001(config-unit-UNIT2)#group interface 2_IG
vEOS-LF001(config-unit-UNIT2)#show active
maintenance
   unit UNIT2
      group interface IG_1
      group bgp 2_BGP
      group bgp BGP_1
      group bgp 1_BGP
      group interface IG_2
      group interface 1_IG
      group bgp BGP_2
      group interface 2_IG
      profile unit TEST
```

**Comment by ClausHolbechArista** [comment]
I don't think you need to do anything in regards to sorting. Just use the `arista.avd.natural_sort` filter on the list of groups. It will not match whatever logic(?) EOS is using, but it will look readable to the humans.
If you decide to work on this, please also include the `groups:`, so we get a full working feature. Thanks a lot for your work on this.

**Comment by kmueller68** [comment]
@daanvdsanden Please allow me to ask what progress you have made on this?
Since we also need this implemented we might incooperate on the development.

We will furthermore work on the addition of "group interface" configuration, which we need in addition to the custom maintenance config:
```
group interface MM_Int_Group_SFP
 interface Et1-48
 maintenance profile interface downlinks
 exit
```

**Comment by daanvdsanden** [comment]
I issued an PR including the groups. At the moment they are split since they also appear splitted in the configuration (on different places. Can easily be changed to ```groups:``` instead of:
```yaml
bgp_groups_:
  < bgp_group_1 >:
    vrf: < vrf_name >
    neighbors:
    - < neighbor_1 >
    - < neighbor_2 >
    - < peer_group >
    maintenance:
      bgp_profiles:
      - < bgp_profile_1 >

interface_groups:
  < interface_group_1 >:
    interfaces:
      ethernet: < ethernet_interface_range >
      port_channel: < port_channel_interface_range >
      vlan: < vlan_interface_range >
    maintenance:
      bgp_profiles:
      - < bgp_profile_1 >
      interface_profiles:
      - < interface_profile_1 >
      - < interface_profile_2 >
```

The reason I didn't use an array for the interface in the interface group is that it is impossible to flatten the Ethernet interfaces without knowing what kind of devices it is and what kind of linecards there are in a chassis. Since Ethernet2/2 can be lane 2 in a QFSP in port 2 but also a SFP+ in a linecard in slot 2.

**Comment by ClausHolbechArista** [comment]
> The reason I didn't use an array for the interface in the interface group is that it is impossible to flatten the Ethernet interfaces without knowing what kind of devices it is and what kind of linecards there are in a chassis. Since Ethernet2/2 can be lane 2 in a QFSP in port 2 but also a SFP+ in a linecard in slot 2.

I don't follow your point here, so if you could show the problem with some example data, I hope I can understand it better.

We have two "competing" PRs now covering the `groups`. Main difference is the `interfaces` which you compile into a list very similar to the EOS CLI output, but you require a more granular data input to do that. Personally I prefer to more "lazy" method of just passing the string handed to us, since I think it is more user-friendly. I acknowledge that this could potentially produce CLI that is valid but different than "show running", but in this case I think it would be so much easier to use, and more in line with the rest of AVD.

So we are talking about:
```yaml
  < interface_group_1 >:
    interfaces:
      ethernet: < ethernet_interface_range >
      port_channel: < port_channel_interface_range >
      vlan: < vlan_interface_range >
```
vs.
```yaml
  < interface_group_1 >:
    interfaces:
      - < interface_range >
      - < interface_range >
      - < interface_range >
```
Let us get a view from @carlbuchmann and @titom73 on this.

When we have the final data model, we need to sort of which PR will cover what.

**Comment by daanvdsanden** [comment]
I like the fact that the generated config is an exact match of the running config. So when I started I also used the "lazy" method and I thought I could flatten the array the same way as is done with vlans on a trunk port. That was impossible for me since to do this properly I need to know for what kind of device I'm creating the config for. That was the point I was trying to make with the Ethernet2/2 example. It can be two different things which can impact the result when the array is flattened.

E.G. the following happens on a 7050CX switch with QSFPs: 
```
EOS(config-group-if-TEST)#interface ethernet 1/1-4
EOS(config-group-if-TEST)#interface ethernet 2/1-4
EOS(config-group-if-TEST)#show active
group interface TEST
   interface Et1/1-2/4
   exit
```
However when the same commands are given on a arista with linecard where SFP(+) modules inserted in slot 1 and 2, I (assume) it will not get flatten the same, I expect this to happen (assumption, I don't have a 7500 to test this):
```
EOS(config-group-if-TEST)#interface ethernet 1/1-4
EOS(config-group-if-TEST)#interface ethernet 2/1-4
EOS(config-group-if-TEST)#show active
group interface TEST
   interface Et1/1-4,2/1-4
   exit
```
